aixplain/modules/pipeline/designer/base.py

```diff
@@ -0,0 +1,370 @@
+    node: Optional["Node"] = None
+    link_: Optional["Link"] = None
+
+    def __init__(
+        self,
+        code: str,
+        data_type: Optional[DataType] = None,
+        value: Optional[str] = None,
+        node: Optional["Node"] = None,
+        param_type: Optional[ParamType] = None,
+    ):
+        self.code = code
+        self.data_type = data_type
+        self.value = value
+
+        # is subclasses do not set the param type, set it to None
+        self.param_type = getattr(self, "param_type", param_type)
+
+        if node:
+            self.attach_to(node)
+
+    def attach_to(self, node: "Node") -> "Param":
+        """
+        Attach the param to the node.
+        :param node: the node
+        :return: the param
+        """
+        assert not self.node, "Param already attached to a node"
+        assert self.param_type, "Param type not set"
+        if self.param_type == ParamType.INPUT:
+            node.inputs.add_param(self)
+        elif self.param_type == ParamType.OUTPUT:
+            node.outputs.add_param(self)
+        else:
+            raise ValueError("Invalid param type")
+        self.node = node
+        return self
+
+    def link(self, to_param: "Param") -> "Param":
+        """
+        Link the output of the param to the input of another param.
+        :param to_param: the input param
+        :return: the param
+        """
+        assert self.node, "Param not attached to a node"
+        assert to_param.param_type == ParamType.INPUT, "Invalid param type"
+        assert self in self.node.outputs, "Param not registered as output"
+        return to_param.back_link(self)
+
+    def back_link(self, from_param: "Param") -> "Param":
+        """
+        Link the input of the param to the output of another param.
+        :param from_param: the output param
+        :return: the param
+        """
+        assert self.node, "Param not attached to a node"
+        assert from_param.param_type == ParamType.OUTPUT, "Invalid param type"
+        assert self.code in self.node.inputs, "Param not registered as input"
+        link = from_param.node.link(self.node, from_param, self)
+        self.link_ = link
+        from_param.link_ = link
+        return link
+
+    def serialize(self) -> dict:
+        return {
+            "code": self.code,
+            "dataType": self.data_type,
+            "value": self.value,
+        }
+
+
+class InputParam(Param):
+
+    param_type: ParamType = ParamType.INPUT
+    is_required: bool = True
+
+    def __init__(self, *args, is_required: bool = True, **kwargs):
+        super().__init__(*args, **kwargs)
+        self.is_required = is_required
+
+
+class OutputParam(Param):
+
+    param_type: ParamType = ParamType.OUTPUT
+
+
+class Link(Serializable):
+    """
+    Link class, this class will be used to link the output of the node to the
+    input of another node.
+    """
+
+    from_node: "Node"
+    to_node: "Node"
+    from_param: str
```

Comment:
Let's keep the problem of multiple param mappings for a node pair in the radar.

Reply:
This is already handled in serialization phase